src/modules/Checkbox/Checkbox.js

```diff
@@ -163,10 +163,28 @@ export default class Checkbox extends Component {
     this.trySetState({ checked: !checked, indeterminate: false })
   }
 
-  handleClick = () => {
+  handleClick = (e) => {
+    const { onChange } = this.props
+    const { checked, indeterminate } = this.state
+
+    // https://github.com/Semantic-Org/Semantic-UI-React/pull/3351
+    //
+    // We need to avoid duplicate calls to onClick.
+    // We need to handle clicks on the root and bubble clicks from the label and input.
+    // The order of the events needs to match that of a vanilla DOM checkbox.
+    //
+    // The browser calls onClick AFTER onChange since onClick is on the root element.
+    // Because of this, we call onClick in handleChange to preserve the correct order.
+    // In that case, we skip calls to onClick here to avoid duplicate calls.
+    if (onChange) return
```

Comment:
Wouldn't this cause issues in case the user has both `onClick` and `onChange` defined?

Reply:
Digging a bit deeper, this change actually breaks the use of `onClick` like described in #3348 , I've updated the tests on controlled components to warn us accordingly. You will notice that if you revert this commit, the tests will pass.

Reply:
Thanks for the double checking.  I have a couple more updates I'll push, then I think we can get a final on this and merge it.

Reply:
Thinking again, it's a bit misleading to have an empty handler like I had done before, I'll push a new commit removing it completely and adding the explanation close to the onClick invocation.